models/repo.go

```diff
@@ -2099,3 +2111,433 @@ func (repo *Repository) GetForks() ([]*Repository, error) {
+}
+
+func (repo *Repository) LocalRepoPath() string {
+	return path.Join(setting.AppDataPath, "tmp/local-repo", com.ToStr(repo.ID))
+}
+
+// UpdateLocalRepo makes sure the local copy of repository is up-to-date.
+func (repo *Repository) UpdateLocalRepo(branchName string) error {
+	return updateLocalCopy(repo.RepoPath(), repo.LocalRepoPath(), branchName)
+}
+
+// DiscardLocalRepoChanges makes sure the local copy of repository is the same as the source
+func (repo *Repository) DiscardLocalRepoChanges(branchName string) error {
+	return discardLocalRepoChanges(repo.LocalRepoPath(), branchName)
+}
+
+// discardLocalRepoChanges discards local commits make sure
+// it is even to remote branch when local copy exists.
+func discardLocalRepoChanges(localPath string, branch string) error {
+	if !com.IsExist(localPath) {
+		return nil
+	}
+	// No need to check if nothing in the repository.
+	if !git.IsBranchExist(localPath, branch) {
+		return nil
+	}
+	if err := git.ResetHEAD(localPath, true, "origin/"+branch); err != nil {
+		return fmt.Errorf("ResetHEAD: %v", err)
+	}
+	return nil
+}
+
+// CheckoutNewBranch checks out a new branch from the given branch name
+func (repo *Repository) CheckoutNewBranch(oldBranchName, newBranchName string) error {
+	return checkoutNewBranch(repo.RepoPath(), repo.LocalRepoPath(), oldBranchName, newBranchName)
+}
+
+func checkoutNewBranch(repoPath, localPath, oldBranch, newBranch string) error {
+	if !com.IsExist(localPath) {
+		if error := updateLocalCopy(repoPath, localPath, oldBranch); error != nil {
+			return error
+		}
+	}
+	if err := git.Checkout(localPath, git.CheckoutOptions {
+		Branch:  newBranch,
+		OldBranch: oldBranch,
+		Timeout: time.Duration(setting.Git.Timeout.Pull) * time.Second,
+	}); err != nil {
+		return fmt.Errorf("Checkout New Branch: %v", err)
+	}
+	return nil
+}
+
+// updateRepoFile adds new file to repository.
+func (repo *Repository) UpdateRepoFile(doer *User, oldBranchName, branchName, oldTreeName, treeName, content, message string, isNewFile bool) (err error) {
+	repoWorkingPool.CheckIn(com.ToStr(repo.ID))
+	defer repoWorkingPool.CheckOut(com.ToStr(repo.ID))
+
+	if err = repo.DiscardLocalRepoChanges(oldBranchName); err != nil {
+		return fmt.Errorf("discardLocalRepoChanges: %s - %v", oldBranchName, err)
+	} else if err = repo.UpdateLocalRepo(oldBranchName); err != nil {
+		return fmt.Errorf("UpdateLocalRepo: %s - %v", oldBranchName, err)
+	}
+
+	if( oldBranchName != branchName ){
+		if err := repo.CheckoutNewBranch(oldBranchName, branchName); err != nil {
+			return fmt.Errorf("CheckoutNewBranch: %s - %s: %v", oldBranchName, branchName, err)
+		}
+	}
+
+	localPath := repo.LocalRepoPath()
+	filePath := path.Join(localPath, treeName)
+
+	if len(message) == 0 {
+		if isNewFile {
+			message = "Add '" + treeName + "'"
+		} else {
+			message = "Update '" + treeName + "'"
+		}
+	}
+
+	if ! com.IsExist(filepath.Dir(filePath)) {
+		os.MkdirAll(filepath.Dir(filePath), os.ModePerm)
```

Comment:
`os.MkdirAll` will do nothing if directory already exists, so we don't need to check here ourselves.